tls/s2n_tls13_handshake.c

```diff
@@ -182,13 +182,21 @@ int s2n_update_application_traffic_keys(struct s2n_connection *conn, s2n_mode mo
     s2n_tls13_key_blob(app_key, conn->secure->cipher_suite->record_alg->cipher->key_material_size);
 
     /* Derives next generation of traffic key */
+    uint8_t *count = NULL;
     POSIX_GUARD(s2n_tls13_derive_traffic_keys(&keys, &app_secret_update, &app_key, &app_iv));
     if (status == RECEIVING) {
         POSIX_GUARD(conn->secure->cipher_suite->record_alg->cipher->set_decryption_key(old_key, &app_key));
+        count = &conn->recv_key_updated;
     } else {
         POSIX_GUARD(conn->secure->cipher_suite->record_alg->cipher->set_encryption_key(old_key, &app_key));
+        count = &conn->send_key_updated;
     }
 
+    /* Increment the count.
+     * Don't treat overflows as errors-- we only do best-effort reporting.
+     */
+    *count = MIN(UINT8_MAX, *count + 1);
```

Comment:
Your test demonstrates that this works, but if count is a uint8, why doesn't it overflow to 0 before being compared to UINT8_MAX in MIN?

Reply:
Aw, give the compiler a _little_ credit :) Math would be a nightmare if adding two uint8_ts could just overflow like that in an intermediate step. There are a bunch of rules around implicit conversions during math, but the relevant one here is that minimally types are converted to int, and "*count + 1" will never be larger than an int.

I found this pretty clear summary of the conversion rules: https://wiki.sei.cmu.edu/confluence/display/c/INT02-C.+Understand+integer+conversion+rules See "Integer Promotions". As an even shorter summary: unless you're dealing with decimals, mismatched signs, or very large numbers, you're generally fine and only need to worry about your final result.

Reply:
Ahh interesting ok. Thanks!